internal/provider/aws/provider.go

```diff
@@ -135,11 +139,19 @@ func WithAddressID(addressID string) AddressIDFilter {
 
 // awsProvider is an AWS provider implementation for the provider.Provider interface
 type awsProvider struct {
-	ec2 *ec2.Client
+	// We have a mutex because the provider can be used by multiple controllers,
+	// so that we can avoid callers result in cache misses and multiple
+	// calls to AWS API when we could have just made one call.
+	sync.Mutex
+	ec2                    *ec2.Client
+	instancesCache         *cache.Cache
+	securityGroupsCache    *cache.Cache
+	networkInterfacesCache *cache.Cache
+	addressesCache         *cache.Cache
```

Comment:
Here we could have a controller blocked by another one while not querying the same resource. 
Could we have a mutex per cache ?

Reply:
This could also be used to have one mutex per cache key: [k8s.io/utils/keymutex](https://pkg.go.dev/k8s.io/utils/keymutex)